students/hirot_directory/session10/oo_mailroom_hiro/mailroom.py

```diff
@@ -0,0 +1,212 @@
+    #         Donor("Paul Allen", [663.23, 43.87, 1.32]),
+    #         Donor("Mark Zuckerberg", [1663.23, 4300.87, 10432.0]),
+    #         ]
+
+
+def find_donor(name):
+    if name in make_sample_db().sample_db.keys():
+        return name
+    else:
+        return None
+
+    # return self.donor_data.get(Donor.normalize_name(name))
+
+# db = DonorDB(make_sample_db())
+
+
+class DonorDB():
+
+    def __init__(self):
+
+        self.all_donors = {}
+
+        # self.all_donors = [Donor(name, donations)
+        #                    for name, donations in zip(self.donor_names, self.donations)]
+
+
+
+
+    def add_donor(self, name):
+        donor = Donor(name)
+        self.all_donors[name] = donor
+        return donor
+
+
+    def print_list(self):
+
+        print("{msg1: <50}| {msg2: <12}| "
+
+              "{msg3:<10}| {msg4: <12}".format(msg1="Donor Name",
+
+                                               msg2="Total Given",
+
+                                               msg3="Num Gifts",
+
+                                               msg4="Average Gift"))
+        for donor in self.all_donors:
+            for key in donor.donor_dict:
+                print(donor.donor_dict)
+                t = sum(donor.donor_dict[key])
+                n = len(donor.donor_dict[key])
+                a = t / n
+
+        print("{d: <50} ${t: 12.2f}{n: 12d}{a: 14.2f}".format(d=donor, t=t, n=n, a=a))
+
+    def create_report(self):
+        print("{msg1: <50}| {msg2: <12}| "
+
+              "{msg3:<10}| {msg4: <12}".format(msg1="Donor Name",
+
+                                               msg2="Total Given",
+
+                                               msg3="Num Gifts",
+
+                                               msg4="Average Gift"))
+        for d in self.donor_dict:
+            t = sum(self.donor_dict[d])
+            n = len(self.donor_dict[d])
+            a = t / n
+
+            print("{d: <50} ${t: 12.2f}{n: 12d}{a: 14.2f}".format(d=d, t=t, n=n, a=a))
+
+    def send_letters(self):
+        try:
+            for n, d in self.donor_dict.items():
+                name = n.replace(" ", "_")
+                f = open('{0}.txt'.format(name), 'wb')
+                d = sum(d)
+                text = ("Dear Mr./Ms. {},\n\n"
+                        "    Thank you for your kind donation of USD ${}.\n"
+                        "    It will be put to very good use.\n\n"
+                        "          Sincerely,\n"
+                        "            -The Team").format(n, d).encode()
+                f.write(text)
+                f.close()
+
+        except IOError as i:
+            print("Error code is", i)
+        print("letter drafts are in the folder for your review."
+              " Please review before mailing out.")
+
+
+class Donor():
+
+    def __init__(self, donor_name, donations=()):
+        self.name = donor_name.strip()
+        self.donations = list(donations)
+
+    def make_donation(self, new_donation):
+
```

Comment:
you may want to add some code here to check that you have a reasonable donation value -- call flost() onitto makeit a number, and maybe check that it's > 0.0 ...